protocols/raft/src/main/java/io/atomix/protocols/raft/storage/compactor/RaftLogCompactor.java

```diff
@@ -0,0 +1,322 @@
+
+  private static final Duration SNAPSHOT_INTERVAL = Duration.ofSeconds(10);
+  private static final Duration MIN_COMPACT_INTERVAL = Duration.ofSeconds(10);
+
+  private static final int COMPACTION_WINDOW_SIZE = 10;
+  private static final int COMPACTION_TIME_FACTOR = 10;
+  private static final long DEFAULT_COMPACTION_TIME = Duration.ofMinutes(1).toMillis();
+
+  private final RaftContext raft;
+  private final ThreadContext threadContext;
+  private final DescriptiveStatistics compactionStats = new DescriptiveStatistics(COMPACTION_WINDOW_SIZE);
+  private final Random random = new Random();
+  private volatile CompletableFuture<Void> compactFuture;
+  private long lastCompacted;
+
+  public RaftLogCompactor(RaftContext raft, ThreadContext threadContext) {
+    this.raft = raft;
+    this.threadContext = threadContext;
+    scheduleSnapshots();
+  }
+
+  /**
+   * Returns a boolean indicating whether the logs should be compacted.
+   */
+  private boolean shouldCompact() {
+    return raft.getStorage().storageLevel() == StorageLevel.MEMORY
+        || raft.getStorage().dynamicCompaction()
+        || raft.getStorage().statistics().getRemainingDuration().toMillis() / 2 < getCompactionTime()
+        || raft.getStorage().statistics().getUsableSpace() / (double) raft.getStorage().statistics().getTotalSpace() < raft.getStorage().freeDiskBuffer();
+  }
+
+  /**
+   * Schedules a snapshot iteration.
+   */
+  private void scheduleSnapshots() {
+    threadContext.schedule(SNAPSHOT_INTERVAL, () -> snapshotServices(true, false));
+  }
+
+  /**
+   * Compacts Raft logs.
+   *
+   * @return a future to be completed once logs have been compacted
+   */
+  public CompletableFuture<Void> compact() {
+    return snapshotServices(false, true);
+  }
+
+  /**
+   * Takes a snapshot of all services and compacts logs if the server is not under high load or disk needs to be freed.
+   */
+  private synchronized CompletableFuture<Void> snapshotServices(boolean rescheduleAfterCompletion, boolean force) {
+    if (compactFuture != null) {
+      if (rescheduleAfterCompletion) {
+        scheduleSnapshots();
+      }
+      return compactFuture;
+    }
+
+    // If the server is under high load and the log doesn't *need* to be compacted, skip snapshotting.
+    if (raft.getStorage().dynamicCompaction() && !shouldCompact()) {
+      if (rescheduleAfterCompletion) {
+        scheduleSnapshots();
+      }
+      return CompletableFuture.completedFuture(null);
+    }
+
+    long lastApplied = raft.getLastApplied();
+
+    // Only take snapshots if segments can be removed from the log below the lastApplied index.
+    if (raft.getLog().isCompactable(lastApplied) && raft.getLog().getCompactableIndex(lastApplied) > lastCompacted) {
+
+      // If the server is under high load, skip compaction and log a message.
+      if (raft.getStorage().dynamicCompaction() && raft.getLoadMonitor().isUnderHighLoad()) {
```

Comment:
I think this condition is not quite right. The problem is, it can skip compaction while the server is under high load unless it's also checking `shouldCompact()` and `force`